astropy/coordinates/distances.py

```diff
@@ -153,37 +152,33 @@ def __new__(
                 # the log of the distance
                 meanlogval = np.log10(value.value).mean()
                 if meanlogval > 6:
-                    unit = u.Mpc
+                    value <<= u.Mpc
                 elif meanlogval > 3:
-                    unit = u.kpc
+                    value <<= u.kpc
                 elif meanlogval < -3:  # ~200 AU
-                    unit = u.AU
-                else:
-                    unit = u.pc
+                    value <<= u.AU
 
         elif parallax is not None:
             parallax = u.Quantity(parallax, copy=False, subok=True)
-            if unit is None:
```

Comment:
Here I mind a bit less, though at some level, this is more explicit and thus it is clearer what is happening. (Quantity has a fast path also if the unit is set but the same, so it doesn't matter to set it.) But also fine to make the change.

Reply:
It's worth pointing out that the updated code _never_ changes `unit`, which is a nice property.